man/im_flood.3
.TH IM_FLOOD 3 "30 October 1992"
.SH NAME
im_flood, im_flood_blob \- flood a area
.SH SYNOPSIS
.B #include <vips/vips.h>

int im_flood( im, x, y, ink, dout )
.br
IMAGE *im;
.br
int x, y;
.br
PEL *ink;
.br
Rect *dout;

int im_flood_blob( im, x, y, ink, dout )
.br
IMAGE *im;
.br
int x, y;
.br
PEL *ink;
.br
Rect *dout;

.SH DESCRIPTION
.B im_flood()
fills an enclosed area from a starting point, painting ink into 4-way
connected pels whose colour is not equal to ink. 

.B im_flood_blob()
floods with the ink colour, finding pels 4-way connected to the start pel
which are the same colour as the start pel. It is useful for changing the
colour of a blob of pels which all have the same value.

.SH RETURN VALUE
All functions return 0 on success and -1 on error.
.SH SEE ALSO
im_insertplace(3), im_smudge(3).


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/inplace/im_flood.c
/* int im_flood( IMAGE *im, int x, int y, PEL *ink, Rect *dout )
 *
 * Flood fill from point (x,y) with colour ink. Flood up to boundary == ink.
 * Any type, any number of bands, IM_CODING_LABQ too. Returns the bounding box 
 * of the modified pixels in dout, whether it succeeds or not.
 *
 * Currently a rather inefficient pixel-based algorithm, should put something
 * better in, really. Speed isn't likely to be a problem, except for very
 * large images.
 *
 * JC 30/8/97
 *	- VIPSified, cleaned up, from "John Robinson's prog to fill 
 *	  enclosed areas"
 *	- something Kirk gave me, so thanks John 
 * JC 1/10/97
 *	- swapped inner memcmp/cpy for a loop ... faster for small pixels
 * 13/7/02 JC
 *	- im_flood_blob() added
 * 5/12/06
 * 	- im_invalidate() after paint
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Size of a point buffer. We allocate a list of these to hold points we need
 * to visit.
 */
#define PBUFSIZE (1000)

/* An xy position.
 */
typedef struct {
	int x, y;
} Point;

/* A buffer of points, and how many of them have been used. When full, alloc a
 * new buffer, and link it on.
 */
typedef struct _Buffer {
	struct _Buffer *next;
	int n;
	Point points[PBUFSIZE];
} Buffer;

/* Our state.
 */
typedef struct {
	/* Parameters.
	 */
	IMAGE *im;
	int x, y;
	PEL *ink;		/* Copy of ink param */
	Rect *dout;		/* Write dirty here at end */

	/* Derived stuff.
	 */
	PEL *edge;		/* Boundary colour */
	int equal;		/* Fill to == edge, or != edge */
	int ps;			/* sizeof( one pel ) */
	int ls;			/* sizeof( one line ) */
	int left, right;	/* Area will fill within */
	int top, bottom;
	Rect dirty;		/* Bounding box of pixels we have changed */

	/* Two buffers of points which we know need checking.
	 */
	Buffer *buf1;
	Buffer *buf2;
} State;

/* Alloc a new buffer.
 */
static Buffer *
build_buffer( void )
{
	Buffer *buf = IM_NEW( NULL, Buffer );

	if( !buf )
		return( NULL );
	buf->next = NULL;
	buf->n = 0;

	return( buf );
}

/* Free a chain of buffers.
 */
static void
free_buffer( Buffer *buf )
{
	IM_FREE( buf->next );
	im_free( buf );
}

/* Free a state.
 */
static void
free_state( State *st )
{
	/* Write dirty back to caller.
	 */
	if( st->dout )
		*st->dout = st->dirty;

	/* Free our stuff.
	 */
	IM_FREE( st->ink );
	IM_FREEF( free_buffer, st->buf1 );
	IM_FREEF( free_buffer, st->buf2 );
	im_free( st );
}

/* Build a state.
 */
static State *
build_state( IMAGE *im, int x, int y, PEL *ink, Rect *dout )
{
	State *st = IM_NEW( NULL, State );

	if( !st )
		return( NULL );
	st->im = im;
	st->x = x;
	st->y = y;
	st->ink = NULL;
	st->dout = dout;
	st->ps = IM_IMAGE_SIZEOF_PEL( im );
	st->ls = IM_IMAGE_SIZEOF_LINE( im );
	st->buf1 = NULL;
	st->buf2 = NULL;
	st->left = 0;
	st->top = 0;
	st->right = im->Xsize;
	st->bottom = im->Ysize;
	st->dirty.left = x;
	st->dirty.top = y;
	st->dirty.width = 0;
	st->dirty.height = 0;

	if( !(st->ink = (PEL *) im_malloc( NULL, st->ps )) ||
		!(st->edge = (PEL *) im_malloc( NULL, st->ps )) ||
		!(st->buf1 = build_buffer()) ||
		!(st->buf2 = build_buffer()) ) {
		free_state( st );
		return( NULL );
	}
	memcpy( st->ink, ink, st->ps );

	return( st );
}

/* Add xy to buffer, move buffer on on overflow.
 */
#define ADD( BUF, X, Y ) { \
	BUF->points[BUF->n].x = X; \
	BUF->points[BUF->n].y = Y; \
	BUF->n++; \
	if( BUF->n == PBUFSIZE ) { \
		if( !BUF->next ) { \
			if( !(BUF->next = build_buffer()) ) \
				return( -1 ); \
		} \
		BUF = BUF->next; \
		BUF->n = 0; \
	} \
}

/* If point != edge, add it to out.
 */
#define ADDIFNOTEDGE( P, X, Y ) { \
	PEL *p1 = (P); \
 	\
	for( j = 0; j < st->ps; j++ ) \
		if( p1[j] != st->edge[j] ) { \
			ADD( out, X, Y ); \
			break; \
		} \
}

/* If point == edge, add it to out.
 */
#define ADDIFEDGE( P, X, Y ) { \
	PEL *p1 = (P); \
 	\
	for( j = 0; j < st->ps; j++ ) \
		if( p1[j] != st->edge[j] ) \
			break; \
	if( j == st->ps ) \
		ADD( out, X, Y ); \
}

/* Read points to fill from in, write new points to out.
 */
static int
dofill( State *st, Buffer *in, Buffer *out )
{
	int i, j;

	/* Clear output buffer.
	 */
	out->n = 0;

	/* Loop over chain of input buffers.
	 */
	for(;;) {
		/* Loop for this buffer.
		 */
		for( i = 0; i < in->n; i++ ) {
			/* Find this pixel.
			 */
			int x = in->points[i].x;
			int y = in->points[i].y;
			PEL *p = (PEL *) st->im->data + x*st->ps + y*st->ls;

			/* Is it still not fore? May have been set by us
			 * earlier.
			 */
			for( j = 0; j < st->ps; j++ )
				if( p[j] != st->ink[j] )
					break;
			if( j == st->ps )
				continue;

			/* Set this pixel.
			 */
			for( j = 0; j < st->ps; j++ )
				p[j] = st->ink[j];

			/* Changes bb of dirty area?
			 */
			if( x < st->dirty.left ) {
				st->dirty.left -= x;
				st->dirty.width += x;
			}
			else if( x > st->dirty.left + st->dirty.width )
				st->dirty.width += x;

			if( y < st->dirty.top ) {
				st->dirty.top -= y;
				st->dirty.height += y;
			}
			else if( y > st->dirty.top + st->dirty.height )
				st->dirty.height += y;

			/* Propogate to neighbours.
			 */
			if( st->equal ) {
				if( x < st->right - 1 )
					ADDIFEDGE( p + st->ps, x + 1, y );
				if( x > st->left )
					ADDIFEDGE( p - st->ps, x - 1, y );
				if( y < st->bottom - 1 )
					ADDIFEDGE( p + st->ls, x, y + 1 );
				if( y > st->top )
					ADDIFEDGE( p - st->ls, x, y - 1 );
			}
			else {
				if( x < st->right - 1 )
					ADDIFNOTEDGE( p + st->ps, x + 1, y );
				if( x > st->left )
					ADDIFNOTEDGE( p - st->ps, x - 1, y );
				if( y < st->bottom - 1 )
					ADDIFNOTEDGE( p + st->ls, x, y + 1 );
				if( y > st->top )
					ADDIFNOTEDGE( p - st->ls, x, y - 1 );
			}
		}

		if( in->n == PBUFSIZE )
			/* Buffer full ... must be another one.
			 */
			in = in->next;
		else
			break;
	}

	return( 0 );
}

int
im_flood( IMAGE *im, int x, int y, PEL *ink, Rect *dout )
{
	State *st;
	Buffer *in, *out, *t;
	PEL *p;

	if( im_rwcheck( im ) )
		return( -1 );
	if( im->Coding != IM_CODING_NONE && im->Coding != IM_CODING_LABQ ) {
		im_error( "im_flood", _( "uncoded or IM_CODING_LABQ only" ) );
		return( -1 );
	}
	if( !(st = build_state( im, x, y, ink, dout )) )
		return( -1 );

	/* Test start pixel ... nothing to do?
	 */
	p = (PEL *) im->data + x*st->ps + y*st->ls;
	if( memcmp( p, ink, st->ps ) == 0 ) {
		free_state( st );
		return( 0 );
	}

	/* Flood to != ink.
	 */
	memcpy( st->edge, ink, st->ps );
	st->equal = 0;

	/* Add start pixel to the work buffer, and loop.
	 */
	ADD( st->buf1, x, y )
	for( in = st->buf1, out = st->buf2; 
		in->n > 0; t = in, in = out, out = t )
		if( dofill( st, in, out ) ) {
			free_state( st );
			return( -1 );
		}
	
	free_state( st );

	im_invalidate( im );

	return( 0 );
}

int
im_flood_blob( IMAGE *im, int x, int y, PEL *ink, Rect *dout )
{
	State *st;
	Buffer *in, *out, *t;
	PEL *p;

	if( im_rwcheck( im ) )
		return( -1 );
	if( im->Coding != IM_CODING_NONE && im->Coding != IM_CODING_LABQ ) {
		im_error( "im_flood", _( "uncoded or IM_CODING_LABQ only" ) );
		return( -1 );
	}
	if( !(st = build_state( im, x, y, ink, dout )) )
		return( -1 );

	/* Edge is set by colour of start pixel.
	 */
	p = (PEL *) im->data + x*st->ps + y*st->ls;
	memcpy( st->edge, p, st->ps );
	st->equal = 1;

	/* Add start pixel to the work buffer, and loop.
	 */
	ADD( st->buf1, x, y )
	for( in = st->buf1, out = st->buf2; 
		in->n > 0; t = in, in = out, out = t )
		if( dofill( st, in, out ) ) {
			free_state( st );
			return( -1 );
		}
	
	free_state( st );

	im_invalidate( im );

	return( 0 );
}

/* A Flood blob we can call from nip. Grr! Should be a way to wrap these
 * automatically. Maybe nip could do it if it seems a RW image argument?
 */
int
im_flood_blob_copy( IMAGE *in, IMAGE *out, int x, int y, PEL *ink )
{
	IMAGE *t;

	if( !(t = im_open_local( out, "im_flood_blob_copy", "t" )) ||
		im_copy( in, t ) ||
		im_flood_blob( t, x, y, ink, NULL ) ||
		im_copy( t, out ) ) 
		return( -1 );

	return( 0 );
}


